internal_types/src/freezable.rs

```
@@ -64,49 +64,128 @@ use std::sync::{Arc, Weak};
 /// }
 /// ```
 ///
+///
+/// A freeze handle can also be acquired asynchronously
+///
+/// ```
+/// use internal_types::freezable::Freezable;
+/// use std::sync::RwLock;
+///
+/// let rt = tokio::runtime::Builder::new_current_thread().build().unwrap();
+///
+/// rt.block_on(async move {
+///     let lockable = RwLock::new(Freezable::new(23));
+///     let fut_handle = lockable.read().unwrap().freeze();
+///
+///     // NB: Only frozen once future resolved to FreezeHandle
+///     *lockable.write().unwrap().get_mut().unwrap() = 56;
+///
+///     let handle = fut_handle.await;
+///
+///     // The contained data now cannot be modified
+///     assert!(lockable.write().unwrap().get_mut().is_none());
+///     // But it can still be read
+///     assert_eq!(**lockable.read().unwrap(), 56);
+///
+///     // --------------
+///     // Do async work
+///     // --------------
+///
+///     // Finish transaction
+///     *lockable.write().unwrap().unfreeze(handle) = 57;
+/// });
+/// ```
+///
 #[derive(Debug)]
-pub struct Freezable<T>(Arc<T>);
+pub struct Freezable<T> {
+    lock: Arc<tokio::sync::Mutex<()>>,
+    payload: T,
+}
 
 impl<T> Freezable<T> {
     pub fn new(payload: T) -> Self {
-        Self(Arc::new(payload))
+        Self {
+            lock: Default::default(),
+            payload,
+        }
     }
 
     /// Returns a `FreezeHandle` that prevents modification
     /// of the contents of `Freezable` until it is dropped
     ///
     /// Returns None if the object is already frozen
-    pub fn try_freeze(&mut self) -> Option<FreezeHandle<T>> {
-        // Verify exclusive
-        self.get_mut()?;
-        Some(FreezeHandle(Arc::downgrade(&self.0)))
+    pub fn try_freeze(&self) -> Option<FreezeHandle> {
+        let guard = Arc::clone(&self.lock).try_lock_owned().ok()?;
```

I always get confused when I see `.ok()?` and have to mentally remind myself that it returns the `None` if `is_none`, rather than silently ignoring the error